File: backtracking/LC17.java
import java.util.*;

public class LC17
{
	public static void main(String[] args)
	{
		System.out.println(letterCombinations("23"));
	}

	// 17. Letter Combinations of a Phone Number	
	static List<String> letterCombinations(String digits)
	{
		String[] map = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
		List<String> list = new ArrayList<String>();
		if (digits == null || digits.isEmpty())
			return list;
		int[] start = new int[1];
		start[0] = 0;
		String next = map[digits.charAt(0) - 48];
		for (int i = 0; i < next.length(); i++)
		{
			StringBuilder combo = new StringBuilder();
			combo.append(next.charAt(i));
			start[0]++;
			recursive(list, combo, digits, map, start);
		}
		return list;
	}

	static void recursive(List<String> list, StringBuilder combo, String digits, String[] map, int[] start)
	{
		if (combo.length() == 0)
			return ;
		if (combo.length() == digits.length())
		{
			list.add(new String(combo));
			combo.deleteCharAt(combo.length() - 1);
			start[0]--;
			return ;
		}

		String next = map[digits.charAt(start[0]) - 48];
		for (int i = 0; i < next.length(); i++)
		{
			combo.append(next.charAt(i));
			start[0]++;
			recursive(list, combo, digits, map, start);
		}
		
		combo.deleteCharAt(combo.length() - 1);
		start[0]--;	
	}
}

File: python/recursion/lc17.py
"""17. Letter Combinations of a Phone Number

Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.



Example:

Input: "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:

Although the above answer is in lexicographical order, your answer could be in any order you want.
"""
class Solution(object):
  def letterCombinations(self, digits):
    """
    :type digits: str
    :rtype: List[str]
    """
    if not digits:
      return []
    
    # The idea: straightforward to use Recursion
    
    # This is a typical recursion problem. The size of the problem is characterized by
    # the length of the string `digits`.
    
    # The Recurion function always returns a list of strings.
    
    # Base case:
    #
    # when `digits` is empty, return a list of one empty string
    
    # when `len(digits)` >= 1, we list out the characters represented by `digits[0]`:
    #  
    # a1, a2, a3,
    
    # and recursively find the solutions for input `digits[1:]`:
    # 
    # s1, s2, s3, ..., sn
    
    # We simply enumerate all the combination between 
    
    # a1, a2, a3
    #
    # and 
    # 
    # s1, s2, s3, ..., sn
    d2c = {
      '2': 'abc',
      '3': 'def',
      '4': 'ghi',
      '5': 'jkl',
      '6': 'mno',
      '7': 'pqrs',
      '8': 'tuv',
      '9': 'wxyz'
    }
    return self.recursion(digits, d2c)  
      
      
  def recursion(self, digits, d2c):      
    if len(digits) == 0:
      return ['']
    
    letters = d2c[digits[0]]
    combinations = self.recursion(digits[1:], d2c)
    
    result = []
    for char in letters:
      for comb in combinations:
        result.append(char + comb)
    return result    
